src/dynmem.hpp

```diff
@@ -15,110 +15,191 @@
+        CudaSafeCall(cudaMemcpy(mem.ptr_d, other.mem.ptr_d, num_elem * sizeof(T), cudaMemcpyDeviceToDevice));
+#endif /* CUDA_TRAD */
     }
     DynMem_(DynMem_ &&other) : num_elem(other.num_elem)
     {
-        ptr_h = other.ptr_h;
-        other.ptr_h = nullptr;
+        mem.ptr_h = other.mem.ptr_h;
+        other.mem.ptr_h = nullptr;
 #ifdef CUFFT
-        ptr_d = other.ptr_d;
-        other.ptr_d = nullptr;
-#endif
+        mem.ptr_d = other.mem.ptr_d;
+        other.mem.ptr_d = nullptr;
+#ifdef CUDA_TRAD
+        mem.currLoc = other.mem.currLoc;
+#endif /* CUDA_TRAD */
+#endif /* CUFFT */
     }
     ~DynMem_()
     {
         release();
     }
-    T *hostMem() { return ptr_h; }
-    const T *hostMem() const { return ptr_h; }
+    T *hostMem() { 
+#ifdef CUDA_TRAD
+        toHost();
+#endif /* CUDA_TRAD */
+        return mem.ptr_h;
+    }
+    const T *hostMem() const { 
+#ifdef CUDA_TRAD
+        toHost();
+#endif /* CUDA_TRAD */
+        return mem.ptr_h;
+    }
 #ifdef CUFFT
-    T *deviceMem() { return ptr_d; }
-    const T *deviceMem() const { return ptr_d; }
-#endif
+    T *deviceMem() {
+#ifdef CUDA_TRAD
+        toDev();
+#endif /* CUDA_TRAD */
+        return mem.ptr_d;
+    }
+    const T *deviceMem() const {
+#ifdef CUDA_TRAD
+        toDev();
+#endif /* CUDA_TRAD */
+        return mem.ptr_d;
+    }
+#endif /* CUFFT */
     void operator=(DynMem_ &rhs) {
         assert(num_elem == rhs.num_elem);
-        memcpy(ptr_h, rhs.ptr_h, num_elem * sizeof(T));
+        memcpy(mem.ptr_h, rhs.mem.ptr_h, num_elem * sizeof(T));
+#ifdef CUDA_TRAD
+        mem.currLoc = rhs.mem.currLoc;
+        CudaSafeCall(cudaMemcpyAsync(mem.ptr_d, rhs.mem.ptr_d, num_elem * sizeof(T), cudaMemcpyDeviceToDevice));
+#endif /* CUDA_TRAD */
     }
     void operator=(DynMem_ &&rhs)
     {
         assert(num_elem == rhs.num_elem);
         release();
-        ptr_h = rhs.ptr_h;
-        rhs.ptr_h = nullptr;
+        mem.ptr_h = rhs.mem.ptr_h;
+        rhs.mem.ptr_h = nullptr;
 #ifdef CUFFT
-        ptr_d = rhs.ptr_d;
-        rhs.ptr_d = nullptr;
-#endif
+        mem.ptr_d = rhs.mem.ptr_d;
+        rhs.mem.ptr_d = nullptr;
+#ifdef CUDA_TRAD
+        mem.currLoc = rhs.mem.currLoc;
+#endif /* CUDA_TRAD */
+#endif /* CUFFT */
+    }
+    T operator[](uint i) const {
+#ifdef CUDA_TRAD
+        toHost();
+#endif /* CUDA_TRAD */
+        return mem.ptr_h[i];
     }
-    T operator[](uint i) const { return ptr_h[i]; }
 private:
     void release()
     {
-#ifdef CUFFT
-        if (ptr_h)
-            mmng.put(ptr_h, num_elem);
-        //CudaSafeCall(cudaFreeHost(ptr_h));
-#else
-        delete[] ptr_h;
-#endif
+        if (mem.ptr_h)
+            mmng.put(reinterpret_cast<struct MemNode<void>*>(&mem), num_elem);
+    }
+
+#ifdef CUDA_TRAD
+    void toHost() const {
+        if (mem.currLoc == CURR_DEV)
+        {
+            mem.currLoc = CURR_HOST;
+            CudaSafeCall(cudaMemcpyAsync(mem.ptr_h, mem.ptr_d, num_elem * sizeof(T), cudaMemcpyDeviceToHost));
```

Comment:
This should be synchronous. Maybe, you also need cudaStreamSynchronize.

Also add `STREAM_PER_THREAD`.